0003-secret-seed-storage-and-key-derivation.md

```diff
@@ -0,0 +1,201 @@
+
+I think this should be optional.
+A COMIT node doesn't hold funds long term.
+It can only control the funds during the execution of a protocol.
+This is different to LND where if someone gets access to your seed they can take all your funds regardless of whether you have channels open or not.
+
+The best attack vector against an unencrypted seed is for the attacker to get short term access to the system COMIT is running on, covertly steal their secret seed and then wait for them to execute a swap and then redeem to their own address.
+Encrypting the seed does give you protection against this.
+
+Having said that, giving an attacker access to your COMIT node will probably give them access to the HTTP API which they can use to redeem funds to their own address.
+This is probably a better attack vector than the above.
+
+Note: We could prevent this attack by giving the node a way of generating the final redeem addresses without getting them from the api.
+
+For the sake of expediency I propose we incrementally increase the security, i.e. (i) leave the secret unencrypted for now, (ii) make encryption an option later on and (iii) make encryption the default eventually.
+
+### Where to store it?
+
+I don't think we should store secret seed in the database file (this is what lnd does).
+This lets users backup the secret seed separately from the database (although they need to backup both).
+Given that the two are linked, we should store a hash of the secret seed along with the swap id (in the same row or whatever).
+This allows the user to change the secret seed while not invalidating the entire database.
+
+By default I'd put it somewhere like:
+
+```
+~/.comit/secret_seed
+```
+
+But its location should be configurable with `secret_seed_file` in the `.toml` config.
+
+The format of the file should be toml with a single entry `secret_seed` (so we can `encrypted_secret_seed` later).
+
+We should make the permissions restrictive so only the user has read permission on it.
+
+#### What about encrypting the database?
+
+I don't think it's worth encrypting the database yet until we have a motivation to do it.
+An attacker stealing the database would learn what protocol executions you're involved in so atm it's just a privacy issue.
+
+We *could* encrypt the database with the `secret_seed` seeing as they're separated.
+
+### How to generate it?
+
+If `secret_seed_file` isn't set, I think we should just generate it from `OsRng` to `~/.comit/secret_seed` if it doesn't exist.
+
+We can make a tool to generate/restore it from a mnemonic and encrypt/decrypt it later on.
+
+### Should it have a checksum?
+
+This probably won't catch much but it seems like a decent idea to attach a checksum to it if it's stored or transferred in a way that could be corrupted (or perhaps corrupted accidentally via a text editor).
+
+## Secret/NodeID Derivation
+
+I think the way we are doing it now is roughly the right way but I recommend modifying it in the following ways:
```

Comment:
I expect the `NodeID` to be the `libp2p id`. What relation/synergy do you propose between both? How is the `libp2p id` set?

Reply:
see answer to @bonomat below.